Reboot - Fix command not found, add Apline support, fix Solaris command

samdoran opened the pull request:
##### SUMMARY
Fixes #46723 - Run `setup` to gather distribution information. Use that to run the correct command for Alpine.
Fixes #47131 - Run `find` module to search common paths to get the full path to the executable since `secure_path` is not always set on the managed node.
Fixes #48986  - Add arguments for older version of Solaris

Supersedes  #45656


##### ISSUE TYPE
<!--- Pick one below and delete the rest -->
- Bugfix Pull Request


##### COMPONENT NAME
`reboot.py`

##### ADDITIONAL INFORMATION
This may need to be broken up into separate pull requests. It was easier to fix these all at once rather than piecemeal, but I can break them up if that's better.

samdoran committed:
Fix various bugs related in reboot

- Use format strings for consistency and improve debug log messages
- Use local variables instead of class attributes in order to be thread safe
- Run setup module to get distribution and version
- Run find module to get full path of shutdown command
- Use ansible_os_family and ansible_distribution to find commands and args
- Use same command for all Solaris/SunOS distributions
- Move delay calculations to properties
- Reliably check for module run failure
- Fix bug in run_test_command() that accidentally made the method work properly
- Use better exceptions rather than Exception
- Use dict literals rather than constructors
- Correct _check_delay() so it always returns a value, not None
- Don't store and return result in run_test_command() because it's not used anywhere

samdoran committed:
Add changelog

abadger commented on lib/ansible/plugins/action/reboot.py:
```diff
@@ -34,81 +34,130 @@ class ActionModule(ActionBase):
+        if value < 0:
+            return value
+        else:
+            return value
```
I think there's a logic error here.  Perhaps if value < 0: return 0?

samdoran replied:
Yup! Should be `0` if less than `0`.

abadger commented on lib/ansible/plugins/action/reboot.py:
```diff
@@ -34,81 +34,130 @@ class ActionModule(ActionBase):
+                        gather_subset='min'))
+                if not module_output['ansible_facts'] or module_output['failed']:
+                    raise AnsibleError('Failed to determine system distribution.  {0}, {1}'.format(module_output['module_stdout'].strip(),
+                                                                                                   module_output['module_stderr'].strip()))
```
You'll need to_native here because module_stdout and module_stderr are always text type. (and will need to be bytes type on python2)

abadger commented on lib/ansible/plugins/action/reboot.py:
```diff
@@ -34,81 +34,130 @@ class ActionModule(ActionBase):
+    @property
+    def distribution(self):
+        if self._distribution is None:
+            self._distribution = dict()
```
MINOR: ```self._distribution = {}``` is more idiomatic (and literals are quicker to evaluate than function calls but that's not going to be a problem here)

samdoran replied:
Awesome! I had that same feedback from @webknjaz this morning and was looking for a good explanation. Now I know.

abadger commented on lib/ansible/plugins/action/reboot.py:
```diff
@@ -34,81 +34,130 @@ class ActionModule(ActionBase):
+                if not module_output['ansible_facts'] or module_output['failed']:
+                    raise AnsibleError('Failed to determine system distribution.  {0}, {1}'.format(module_output['module_stdout'].strip(),
+                                                                                                   module_output['module_stderr'].strip()))
+            except KeyError as ke:
```
This exception might be a little too tricky... If I understand it, it's handling the non-error case because you provoke KeyError if everything succeeded.  We shouldn't do that except in really special cases :-)

samdoran replied:
Agreed. I was stuck trying to test for if the run failed, but the `failed` key is only present when failed. I don't get a `failed: False` on success.

I toyed with using `defaultdict`, but that seemed too complicated too.

I'm open to suggestions on how to test if the module run was successful.

samdoran replied:
Maybe it'd be better to test `if 'failed' in module_output.keys()`? Wish I'd thought of that earlier.

abadger replied:
yeah, or even simpler:
``` python
if 'failed' in module_output:
    raise AnsibleError([...])
```

containment checks (in) for dicts use the dict's keys already.

abadger commented on lib/ansible/plugins/action/reboot.py:
```diff
@@ -34,81 +34,130 @@ class ActionModule(ActionBase):
+                    patterns=[shutdown_bin],
+                    file_type='any'))
+            full_path = [x['path'] for x in find_result['files']]
+            if len(full_path) == 0:
```
any ```if len(x) == 0:``` can instead be ```if not x:```

abadger commented on lib/ansible/plugins/action/reboot.py:
```diff
@@ -265,6 +311,7 @@ def validate_reboot(self):
     def run(self, tmp=None, task_vars=None):
         self._supports_check_mode = True
         self._supports_async = True
+        self._task_vars = task_vars
```
This never feels right to me.  If we think of task_vars as being global to the instance, then it should be set externally, either in the constructor or by the caller setting it explicitly.  If we think of task_vars as being part of this method, then we shouldn't be setting the instance variable with it.

samdoran replied:
I revamped everything to use local variables rather than storing `self._task_vars` as a class attribute.

abadger commented on lib/ansible/plugins/action/reboot.py:
```diff
@@ -34,86 +34,148 @@ class ActionModule(ActionBase):
+    def distribution(self):
+        if self._distribution:
+            return self._distribution
+        else:
```
```suggestion

```
You don't need the else anymore since the condition is going to return.

samdoran replied:
Removed.

jborean93 commented on lib/ansible/plugins/action/reboot.py:
```diff
@@ -67,53 +65,104 @@ def __init__(self, *args, **kwargs):
+            module_name='setup',
+            module_args=dict(gather_subset='min'))
+        try:
+            if 'failed' in module_output.keys():
```
Can `failed=False` be set in this? Sounds like you would want `if module_output.get('failed', False):` instead?

samdoran replied:
This was tricky and I couldn't nail down a good way to test for this. The `failed` key only exists if there is a failure (in general), but you are correct that `failed=False` could happen.

I think your solution is what I was looking for. 😄

jborean93 commented on lib/ansible/plugins/action/reboot.py:
```diff
@@ -67,53 +65,104 @@ def __init__(self, *args, **kwargs):
+                    to_native(module_output['module_stdout']).strip(),
+                    to_native(module_output['module_stderr']).strip()))
+            distribution['name'] = module_output['ansible_facts']['ansible_distribution'].lower()
+            distribution['version'] = str(module_output['ansible_facts']['ansible_distribution_version'].split('.')[0])
```
Should you be using `to_text` here and the other places you are using `str()`?

samdoran replied:
Yes! Bad habit.

jborean93 commented on lib/ansible/plugins/action/win_reboot.py:
```diff
@@ -66,7 +69,7 @@ def perform_reboot(self):
 
             # Initiate reboot again
-            result2 = self._low_level_execute_command('shutdown /r /t %d' % pre_reboot_delay, sudoable=self.DEFAULT_SUDOABLE)
+            result2 = self._low_level_execute_command('shutdown /r /t {0}'.format(self.pre_reboot_delay), sudoable=self.DEFAULT_SUDOABLE)
```
Should we be using `reboot_command` again. I think the only difference between the current and that is the latter has a comment added which is ok to use here.

samdoran replied:
Seems reasonable. I wasn't sure if attempt number two needed a specifically different command for some reason. I'll use `reboot_command` here.

jborean93 commented on lib/ansible/plugins/action/reboot.py:
```diff
@@ -67,53 +65,104 @@ def __init__(self, *args, **kwargs):
+                    action=self._task.action))
+
+    def get_system_boot_time(self, distribution):
+        boot_time_command = self._get_value_from_facts('BOOT_TIME_COMMANDS', distribution, 'DEFAULT_BOOT_TIME_COMMAND')
```
This is failing on Windows with `"msg": "family"` as `family` isn't defined on `get_distribution` for Windows.

samdoran replied:
My Virtualbox environment got messed up and I wasn't able to reach the Windows VM anymore. I think I finally got it fixed today. Thanks for tracking this down.

jborean93 commented on lib/ansible/plugins/action/reboot.py:
```diff
@@ -187,69 +242,86 @@ def do_until_success_or_timeout(self, action, reboot_timeout, action_desc):
 
             if connect_timeout:
                 try:
                     self._connection.set_option("connection_timeout", connect_timeout)
```
I know this isn't part of the changes but I just found this. `connect_timeout` is the value of `connect_timeout` or `connect_timeout_sec` as part of the reboot task. I believe we want to use `self._original_connection_timeout` if it has been set instead.

samdoran replied:
Good catch. I will fix this.

webknjaz commented on lib/ansible/plugins/action/reboot.py:
```diff
@@ -34,110 +34,160 @@ class ActionModule(ActionBase):
+    def _check_delay(self, key):
+        value = int(self._task.args.get(key, self._task.args.get(key + '_sec', self.DEFAULT_POST_REBOOT_DELAY)))
+        if value < 0:
+            value = 0
```
You return `None` in this case. Fix it:
```suggestion
            return 0
```

webknjaz replied:
(or drop "else" below)

samdoran replied:
I added a test case for this.

webknjaz commented on lib/ansible/plugins/action/reboot.py:
```diff
@@ -34,110 +34,160 @@ class ActionModule(ActionBase):
+        return self._check_delay('post_reboot_delay')
 
+    def _check_delay(self, key):
+        value = int(self._task.args.get(key, self._task.args.get(key + '_sec', self.DEFAULT_POST_REBOOT_DELAY)))
```
you use `DEFAULT_POST_REBOOT_DELAY` for both pre-/post-
maybe rename it?

samdoran replied:
Good catch. I'll add a `default` param to the method.

webknjaz commented on lib/ansible/plugins/action/reboot.py:
```diff
@@ -34,110 +34,160 @@ class ActionModule(ActionBase):
+            return value
+
+    def _get_value_from_facts(self, variable_name, distribution, default_value):
+        ''' Get dist+version specific args first, then distribution, then family, lastly use default '''
```
I prefer `""""` and a sentence specified by PEP257

samdoran replied:
Thank you for that info. I'll use double quotes from now on.

webknjaz commented on lib/ansible/plugins/action/reboot.py:
```diff
@@ -34,110 +34,160 @@ class ActionModule(ActionBase):
+        find_result = self._execute_module(
+            task_vars=task_vars,
+            module_name='find',
+            module_args=dict(
```
Please use dict literals.

samdoran replied:
I did some reading and playing with `timeit` and `dis` to compare these methods a bit. I think I better understand when to use literals or the constructor now. This is a habit I picked up from most of our modules which use constructors.

webknjaz commented on lib/ansible/plugins/action/reboot.py:
```diff
@@ -150,25 +200,33 @@ def run_test_command(self, **kwargs):
             result['failed'] = True
-            result['msg'] = 'test command failed: %s %s' % (to_native(command_result['stderr'], to_native(command_result['stdout'])))
+            result['msg'] = msg
+            raise Exception(msg)
```
Please raise some semantically correct exception, like
```suggestion
            raise RuntimeError(msg)
```

Also, args are messed up: you construct `result` dict in this if-branch but not use it at all?

samdoran replied:
I used more appropriate exceptions when raising.

samdoran replied:
I noticed this yesterday, too. It seems I was originally planning to return the output of the test command and do something with it, but never did. Therefore there's no need to store any of this.

samdoran committed:
Improve tests

- add test for post reboot command that fails
- test negative values for delay parameters

maxamillion commented:
shipit
